src/main/java/com/xceptance/common/io/Utf8Reader.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2005-2023 Xceptance Software Technologies GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.xceptance.common.io;
+
+import java.io.CharArrayReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * A {@link Reader} implementation that reads UTF-8-encoded text from an {@link InputStream} throwing an
+ * {@link IOException} if the text is not valid UTF-8.
+ * <p>
+ * Note: This class buffers the content of the stream in memory so it should be used for small amounts of data only.
+ */
+public class Utf8Reader extends Reader
+{
+    /**
+     * The reader we delegate to when actually dealing out the characters.
+     */
+    private final CharArrayReader charArrayReader;
+
+    /**
+     * Creates a new {@link Utf8Reader} instance.
+     * 
+     * @param inputStream
+     *            the stream to read the text from
+     * @throws IOException
+     *             if the input stream could not be read or does not represent UTF-8-encoded text
+     */
+    public Utf8Reader(final InputStream inputStream) throws IOException
+    {
+        final byte[] bytes = inputStream.readAllBytes();
+        char[] chars;
+
+        try
+        {
+            chars = getAsChars(bytes, StandardCharsets.UTF_8);
+        }
+        catch (final CharacterCodingException cce)
+        {
+            throw new IOException("Data does not represent UTF-8-encoded text", cce);
+        }
+
+        charArrayReader = new CharArrayReader(chars);
```

Comment:
Instance field is only initialized when reading all bytes did succeed. IMO, the following makes the control flow more clear:
```suggestion
        
        try
        {
            final char[] chars = getAsChars(bytes, StandardCharsets.UTF_8);
            charArrayReader = new CharArrayReader(chars);
        }
        catch (final CharacterCodingException cce)
        {
            throw new IOException("Data does not represent UTF-8-encoded text", cce);
        }
```